enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,617 @@
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=256
+    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
+    // +optional
+    Name *string `json:"name"`
+
+    // Set defines HTTP headers that should be set or replaced or appended.
+    // If this field is empty, no headers are added.
+    //
+    // Note that this option only applies to cleartext HTTP connections
+    // and to secure HTTP connections for which the ingress controller
+    // terminates encryption (that is, edge-terminated or reencrypt
+    // connections).  Headers cannot be set(replace)/append for TLS passthrough
+    // connections.
+    //
+    // +kubebuilder:validation:Optional
+    // +optional
+    Set *IngressControllerSetHTTPHeader `json:"set,omitempty"`
+
+    // delete defines HTTP headers that should be deleted.
+    // If this field is empty, no headers are deleted.
+    //
+    // Note that this option only applies to cleartext HTTP connections
+    // and to secure HTTP connections for which the ingress controller
+    // terminates encryption (that is, edge-terminated or reencrypt
+    // connections).  Headers cannot be set(replace)/append for TLS passthrough
+    // connections.
+    //
+    // +kubebuilder:validation:Optional
+    // +optional
+    Delete *IngressControllerDeleteHTTPHeader `json:"delete,omitempty"`
+}
+```
+
+```go
+// IngressControllerSetAction is used to understand as whether you want to perform Set related 
+// operation like Replace/Append Or a Delete operation
+// +kubebuilder:validation:Enum=Set;Delete
+type IngressControllerSetHTTPHeaderAction string
+
+const (
+    Set    IngressControllerSetHTTPHeaderAction = "Set"
+    Delete IngressControllerSetHTTPHeaderAction = "Delete"
+)
+```
+
+The IngressControllerSetHTTPHeader type has fields for specifying a header name, header value and action.
+```go
+type IngressControllerSetHTTPHeader struct{
+	// Value specifies a header value.
+    // +kubebuilder:validation:Required 
+    // +required
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=4096
+    Value string `json:"value"`
+
+    // Action specifies a header value.
+    // +kubebuilder:validation:Required
+    // +required
+    // +kubebuilder:validation:Enum=Append;Replace
+    Action IngressControllerHTTPHeaderAction `json:"action"`
+}
+```
+
+This defines Actions to be perfomed on headers when action field is `Set` in the IngressControllerHTTPHeaders. 
+```go
+// IngressControllerHTTPHeaderAction is a policy for setting HTTP headers.
+// +kubebuilder:validation:Enum=Append;Replace
+type IngressControllerHTTPHeaderAction string
+
+const (
+    // AppendHTTPHeader appends the header, preserving any existing header.
+    AppendHTTPHeader IngressControllerHTTPHeaderAction = "Append"
+    // ReplaceHTTPHeader sets the header, removing any existing header.
+    ReplaceHTTPHeader IngressControllerHTTPHeaderAction = "Replace"
```

Comment:
I still think that the name is a little misleading for the end user. `http-response set-header`which will be used for the replace action can not only replace a header but can add the header if it doesn't exist. We need to do something about it, either change the action name to reflect better what will happen or change the implementation.
I personally thought (maybe too radical though) of getting rid of the action type as a whole: this would map 1-1 with `http-response set-header` config command.

Reply:
It would be helpful to have a table in the implementation details showing how each action maps to a stanza in `haproxy.config`.

Reply:
As discussed during the design meeting (09/01/23): no explicit requirement about adding header or replacing by regexp was mentioned in RFE. We assume `http-response set-header` (add if doesn't exist replace if does) stanza is what corresponds the most to the customer expectations.

Reply:
done